examples/mnist-classification/src/mnist_classification/core.clj

```diff
@@ -0,0 +1,214 @@
+
+(defn- image-aug-pipeline
+  "Uses think.image augmentation to vary training inputs."
+  [image]
+  (-> image
+      (image-aug/rotate (- (rand-int (* 2 max-image-rotation-degrees))
+                           max-image-rotation-degrees)
+                        false)
+      (image-aug/inject-noise (* 0.25 (rand)))))
+
+
+
+
+(defn- mnist-observation->image
+  "Creates a BufferedImage suitable for web display from the raw data
+  that the net expects."
+  [observation]
+  (patch/patch->image observation image-size))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; The classification experiment system needs a way to go back and forth from
+;; softmax indexes to string class names.
+(def class-mapping
+  {:class-name->index (zipmap (map str (range 10)) (range))
+   :index->class-name (zipmap (range) (map str (range 10)))})
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Main entry point. In general, a classification experiment trains a net
+;; forever, providing live updates on a local web server.
+(defn train-forever
+  ([] (train-forever {}))
+  ([argmap]
+   (ensure-images-on-disk!)
+   (println "Training forever.")
+   (let [training-folder (str dataset-folder "training")
+         test-folder (str dataset-folder "test")
+         [train-ds test-ds] [(-> training-folder
+                                 (experiment-util/create-dataset-from-folder :image-aug-fn (:image-aug-fn argmap))
+                                 (experiment-util/infinite-class-balanced-dataset))
+                             (-> test-folder
+                                 (experiment-util/create-dataset-from-folder :image-aug-fn (:image-aug-fn argmap)))]]
+     (classification/perform-experiment initial-description
+                                        train-ds test-ds
+                                        mnist-observation->image
+                                        class-mapping
+                                        (merge argmap
+                                               (:image-aug-fn image-aug-pipeline))))))
+
+
+(defn train-forever-uberjar
+  ([] (train-forever-uberjar {}))
+  ([argmap]
+   (println "Training forever from uberjar.")
+   (train-forever argmap)))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Once a net is trained (and the trained model is saved to a nippy file), it
+;; is hopefully straight forward to use that saved model to make inferences on
+;; additional observations. Note that nothing in this section depends on
+;; `experiment`, only `cortex` itself. This makes deploying traned models much
+;; simpler, since `cortex` has many fewer dependencies than `experiment.`
+(def network-filename
+  (str train/default-network-filestem ".nippy"))
+
+
+(defn label-one
+  "Take an arbitrary test image and label it."
+  []
+  (ensure-images-on-disk!)
+  (let [observation (->> (str dataset-folder "test")
```

Comment:
Something like the following `join-path` function is more robust as it doesn't rely on having a `/` at the end of the directory name, but also doesn't fail if it does (I think?).
```
(defn- join-path
  [& args]
(.toString (apply io/file args)))
```